File: Week_03/kotlin/src/_50_PowerXN.kt
class _50_UsingMath {
    // 时间 O(logn) 空间 O(1)
    fun myPow(x: Double, n: Int): Double {
        var m = if (n < 0) -n.toLong() else n.toLong()
        var y = x
        var ret = 1.0
        while (m != 0L) {
            if (m and 1 == 1L) {
                ret *= y
            }
            y *= y
            m = m ushr 1
        }
        return if (n < 0) 1 / ret else ret
    }
}

File: Week_03/c/src/_50_PowerXN.c
double dfs(double, long long);
// 时间 O(logn) 空间 O(logn)
double myPow(double x, int n)
{
    if (x == 0)
        return 0;
    long long N = n < 0 ? -(long long)n : n;
    return n < 0 ? 1 / dfs(x, N) : dfs(x, N);
}
double dfs(double x, long long n)
{
    if (n == 0)
        return 1;
    return (n & 1) ? x * dfs(x * x, n >> 1) : dfs(x * x, n >> 1);
}

// 时间 O(logn) 空间 O(1)
double myPowIterative(double x, int n)
{
    if (x == 0)
        return 0;
    long long N = n < 0 ? -(long long)n : n;
    double ret = 1;
    // 假设n的二进制位为bmbm-1...b3b2b1
    // x^n = x^(2^m-1 * bm...2^2b3 * 2^1* b2*2^0*b1) = x
    while (N)
    {
        if (N & 1)
        {
            ret *= x;
        }
        x *= x;
        N >>= 1;
    }
    return n < 0 ? 1 / ret : ret;
}